nitro-protocol/contracts/VirtualMarginApp.sol

```diff
@@ -0,0 +1,149 @@
+
+            require(candidate.variablePart.turnNum >= 2, 'turnNum < 2; |proof|=1');
+
+            require(
+                NitroUtils.isClaimedSignedBy(candidate.signedBy, 0),
+                'redemption not signed by Leader'
+            );
+
+            require(
+                NitroUtils.isClaimedSignedBy(candidate.signedBy, nParticipants - 1),
+                'redemption not signed by Receiver'
+            );
+
+            _requireCorrectOutcomes(
+                proof[0].variablePart.outcome,
+                candidate.variablePart.outcome,
+                fixedPart.participants[0],
+                fixedPart.participants[nParticipants - 1],
+                nParticipants
+            );
+            return;
+        }
+        revert('bad proof length');
+    }
+
+    function _requireProofOfUnanimousConsensusOnPostFund(
+        RecoveredVariablePart memory rVP,
+        uint256 numParticipants
+    ) internal pure {
+        require(rVP.variablePart.turnNum == 1, 'bad proof[0].turnNum; |proof|=1');
+        require(
+            NitroUtils.getClaimedSignersNum(rVP.signedBy) == numParticipants,
+            'postfund !unanimous; |proof|=1'
+        );
+    }
+
+    function _requireCorrectOutcomes(
+        Outcome.SingleAssetExit[] memory oldOutcome,
+        Outcome.SingleAssetExit[] memory newOutcome,
+        address Leader,
+        address Receiver,
+        uint8 nParticipants
+    ) internal pure {
+        // NOTE: do we need such strict rules?
+        // is there a scenario they can be broken in a malicious way?
+
+        // only 1 collateral asset (USDT) for now, 2 later (+ YellowToken)
+        require(oldOutcome.length == 1 && newOutcome.length == 1, 'invalid number of assets');
+
+        // only 2 allocations
+        require(
+            oldOutcome[0].allocations.length == 2 && newOutcome[0].allocations.length == 2,
+            'invalid number of allocations'
+        );
+
+        // TODO: allocations are to Leader and Receiver
+        // require(
+        //     oldOutcome[0].allocations[0].destination == Leader &&
+        //     oldOutcome[0].allocations[1].destination == Receiver &&
+        //     newOutcome[0].allocations[0].destination == Leader &&
+        //     newOutcome[0].allocations[1].destination == Receiver,
+        //     'invalid number of allocations'
+        // );
+
+        // TODO: Add getter and setter, for Fee and collateral currencies
+        // newOutcome[0].asset == ASSET_FEE_ADDRESS &&
+        // newOutcome[1].asset == ASSET_COLLATERAL_ADDRESS,
+
+        // equal sums
+        uint256 oldAllocationSum;
+        uint256 newAllocationSum;
+        for (uint256 i = 0; i < nParticipants; i++) {
+            oldAllocationSum += oldOutcome[0].allocations[i].amount;
+            newAllocationSum += newOutcome[0].allocations[i].amount;
+        }
+        require(oldAllocationSum == newAllocationSum, 'total allocated cannot change');
```

Comment:
@nksazonov don't you think this is true for all channels and NitroAdjudicator should be checking this?

Reply:
No, `NitroAdjudicator` does not check this.
They mention that participants must be careful about outcomes: https://docs.statechannels.org/protocol-tutorial/0060-funding-a-channel/#outcome-priority